src/libguac/client.c

```diff
@@ -478,6 +478,29 @@ int guac_client_load_plugin(guac_client* client, const char* protocol) {
 
 }
 
+int guac_client_owner_send_required(guac_client* client, const char** required) {
+
+    /* Don't send require instruction if client does not support it. */
+    if (!guac_client_supports_require(client))
+        return -1;
+    
+    int retval;
+
+    pthread_rwlock_rdlock(&(client->__users_lock));
+
+    /* Invoke callback with current owner */
+    retval = guac_protocol_send_required(client->__owner->socket, required);
+    
+    /* Flush the socket */
+    guac_socket_flush(client->__owner->socket);
+
+    pthread_rwlock_unlock(&(client->__users_lock));
```

Comment:
Rather than manually lock `__users_lock`, we should leverage `guac_client_for_owner()`.

Reply:
So, the only thing that I'm struggling with here is the callback - the current specification for the callback allows for the one `void *` data field, which I need to use to pass the required parameters through.  I haven't figured out a good way to get the `int` return value of the underlying `guac_protocol_send_required()` function to pass through the `void *` return value of the callback function.

FWIW, I'm not sure how much it really matters to track the return value of `guac_protocol_send_required()` - in the all of the places where it was used directly in the past I don't think I ever made any use of the return value, but it makes me a little twitchy to have a function that returns something that just never gets looked at.

Hints or opinions on the best way to handle this, all-around?

Reply:
Yep - there's a standard C99 way to losslessly transform between `int` and ` void*` and back: `intptr_t`, provided by `stdint.h`.

Here's an example of `int` to `void*`:

https://github.com/apache/guacamole-server/blob/c10ceab7e80f9d6eae9ad98254bf97f41a0de112/src/guaclog/keydef.c#L166

And here's an example of `void*` to `int`:

https://github.com/apache/guacamole-server/blob/c10ceab7e80f9d6eae9ad98254bf97f41a0de112/src/guaclog/keydef.c#L143

If `intptr_t` weren't a possibility, the next option would be to pass everything that the callback needs within a `struct`, including a location to store what would otherwise be its return value.

Reply:
Brilliant.  I think I've managed to implement this correctly with `intptr_t`.